arviz/data/io_pyro.py

```diff
@@ -1,119 +1,131 @@
-    """
-    sample_point = posterior.exec_traces[0]
-    nodes = [node for node in sample_point.nodes.values() if node["type"] == "sample"]
-    observed = [node["name"] for node in nodes if node["is_observed"]]
-    latent = [node["name"] for node in nodes if not node["is_observed"]]
-    return observed, latent
+_log = logging.getLogger(__name__)
 
 
 class PyroConverter:
     """Encapsulate Pyro specific logic."""
 
-    def __init__(self, *, posterior, coords=None, dims=None):
-        """Convert pyro data into an InferenceData object.
+    nchains = None  # type: int
+    ndraws = None  # type: int
+
+    def __init__(
+        self, *, posterior=None, prior=None, posterior_predictive=None, coords=None, dims=None
+    ):
+        """Convert Pyro data into an InferenceData object.
 
         Parameters
         ----------
-        posterior : pyro.MCMC
+        posterior : pyro.infer.MCMC
             Fitted MCMC object from Pyro
+        prior: dict
+            Prior samples from a Pyro model
+        posterior_predictive : dict
+            Posterior predictive samples for the posterior
         coords : dict[str] -> list[str]
             Map of dimensions to coordinates
         dims : dict[str] -> list[str]
             Map variable names to their coordinates
         """
         self.posterior = posterior
-        self.observed_vars, self.latent_vars = _get_var_names(posterior)
+        if posterior is not None:
+            self.nchains, self.ndraws = posterior.num_chains, posterior.num_samples
+        else:
+            self.nchains = self.ndraws = 0
+        self.prior = prior
+        self.posterior_predictive = posterior_predictive
         self.coords = coords
         self.dims = dims
         import pyro
 
         self.pyro = pyro
 
+    @requires("posterior")
     def posterior_to_xarray(self):
         """Convert the posterior to an xarray dataset."""
-        # Do not make pyro a requirement
-        from pyro.infer import EmpiricalMarginal
-
-        try:  # Try pyro>=0.3 release syntax
-            data = {
-                name: utils.expand_dims(samples.enumerate_support().squeeze())
-                if self.posterior.num_chains == 1
-                else samples.enumerate_support().squeeze()
-                for name, samples in self.posterior.marginal(
-                    sites=self.latent_vars
-                ).empirical.items()
-            }
-        except AttributeError:  # Use pyro<0.3 release syntax
-            data = {}
-            for var_name in self.latent_vars:
-                # pylint: disable=no-member
-                samples = EmpiricalMarginal(
-                    self.posterior, sites=var_name
-                ).get_samples_and_weights()[0]
-                samples = samples.numpy().squeeze()
-                data[var_name] = utils.expand_dims(samples)
+        data = self.posterior.get_samples(group_by_chain=True)
+        data = {k: v.detach().cpu().numpy() for k, v in data.items()}
```

Comment:
If only one chain is used, would this work?

Reply:
yes, this only converts pytorch tensor to numpy. The `group_by_chain=True` will make sure that the sample shape is nchains x ndraws x shape, even with 1 chain.